include/grpcpp/impl/codegen/server_callback.h

```diff
@@ -117,79 +191,575 @@ class CallbackUnaryHandler : public MethodHandler {
       : public experimental::ServerCallbackRpcController {
    public:
     void Finish(Status s) override {
-      finish_tag_.Set(
-          call_.call(),
-          [this](bool) {
-            grpc_call* call = call_.call();
-            auto call_requester = std::move(call_requester_);
-            this->~ServerCallbackRpcControllerImpl();  // explicitly call
-                                                       // destructor
-            g_core_codegen_interface->grpc_call_unref(call);
-            call_requester();
-          },
-          &finish_buf_);
+      finish_tag_.Set(call_.call(), [this](bool) { MaybeDone(); },
+                      &finish_ops_);
       if (!ctx_->sent_initial_metadata_) {
-        finish_buf_.SendInitialMetadata(&ctx_->initial_metadata_,
+        finish_ops_.SendInitialMetadata(&ctx_->initial_metadata_,
                                         ctx_->initial_metadata_flags());
         if (ctx_->compression_level_set()) {
-          finish_buf_.set_compression_level(ctx_->compression_level());
+          finish_ops_.set_compression_level(ctx_->compression_level());
         }
         ctx_->sent_initial_metadata_ = true;
       }
       // The response is dropped if the status is not OK.
       if (s.ok()) {
-        finish_buf_.ServerSendStatus(&ctx_->trailing_metadata_,
-                                     finish_buf_.SendMessage(resp_));
+        finish_ops_.ServerSendStatus(&ctx_->trailing_metadata_,
+                                     finish_ops_.SendMessage(resp_));
       } else {
-        finish_buf_.ServerSendStatus(&ctx_->trailing_metadata_, s);
+        finish_ops_.ServerSendStatus(&ctx_->trailing_metadata_, s);
       }
-      finish_buf_.set_core_cq_tag(&finish_tag_);
-      call_.PerformOps(&finish_buf_);
+      finish_ops_.set_core_cq_tag(&finish_tag_);
+      call_.PerformOps(&finish_ops_);
     }
 
     void SendInitialMetadata(std::function<void(bool)> f) override {
       GPR_CODEGEN_ASSERT(!ctx_->sent_initial_metadata_);
-
-      meta_tag_.Set(call_.call(), std::move(f), &meta_buf_);
-      meta_buf_.SendInitialMetadata(&ctx_->initial_metadata_,
+      callbacks_outstanding_++;
+      meta_tag_.Set(call_.call(),
+                    [this, f](bool ok) {
```

Comment:
I don't have as much experience with lambdas.

Could this be a reference or const reference? We are now adding a copy instead of just moving the function like before, right?

Reply:
Reference won't work for lifetime reasons. Move would have been nice, but move lambda is a c++14 feature that we can't use. Anyway, this method is uncommonly used and not in any performance benchmarks that we care about afaik.

Reply:
I'm adding a todo to indicate that we can consider making this a movable capture if we adopt C++14.